lib/raxx/middleware.ex

```diff
@@ -0,0 +1,49 @@
+defmodule Raxx.Middleware do
+  alias Raxx.Server
+
+  @typedoc """
+  The behaviour module and state/config of a raxx middleware
+  """
+  @type t :: {module, state}
+
+  @typedoc """
+  State of middleware.
+  """
+  @type state :: Server.state()
+
+  @type next :: {[Raxx.part()], state, Server.t()}
+
+  @doc """
+  """
+  @callback process_head(Raxx.Request.t(), state(), inner_server :: Server.t()) :: next()
+
+  @doc """
+  Called every time data from the request body is received
+  """
+  @callback process_data(binary(), state(), inner_server :: Server.t()) :: next()
+
+  @doc """
+  Called once when a request finishes.
+
+  This will be called with an empty list of headers is request is completed without trailers.
+  """
+  @callback process_tail([{binary(), binary()}], state(), inner_server :: Server.t()) :: next()
+
+  @doc """
+  Called for all other messages the server may recieve
+  """
+  @callback process_info(any(), state(), inner_server :: Server.t()) :: next()
+
+  @doc false
+  @spec is_implemented?(module) :: boolean
+  def is_implemented?(module) when is_atom(module) do
+    # taken from Raxx.Server
```

Comment:
Can now just use the function in Raxx.Server. It is public but not documented

Reply:
it checks if the Middleware behaviour is implemented, not the Server one, so the implementation is the same, but it does a different thing